SourceX/controls/plrctrls.cpp

```diff
@@ -203,6 +213,75 @@ void CheckMonstersNearby()
 	}
 }
 
+void FindMeleeTarget()
+{
+	bool visited[MAXDUNX][MAXDUNY] = { 0 };
+	int maxSteps = 25;
+	int rotations = 5;
+	int x = plr[myplr]._px;
+	int y = plr[myplr]._py;
+
+	std::list<bfsNode> queue;
+
+	visited[x][y] = true;
+	queue.push_back({ x, y, 0 });
+
+	while (!queue.empty()) {
+		bfsNode node = queue.front();
+		queue.pop_front();
+
+		for (int i = 0; i < 8; i++) {
+			int dx = node.x + pathxdir[i];
+			int dy = node.y + pathydir[i];
+
+			if (visited[dx][dy])
+				continue; // already visisted
+
+			if (!PosOkPlayer(myplr, dx, dy)) {
+				visited[dx][dy] = true;
+
+				if (dMonster[dx][dy] != 0) {
+					int mi = dMonster[dx][dy] > 0 ? dMonster[dx][dy] - 1 : -(dMonster[dx][dy] + 1);
+					if (CanTargetMonster(mi)) {
+						int newRotations = GetRoteryDistance(dx, dy);
+						if (rotations < newRotations || (pcursmonst != -1 && CanTalkToMonst(i)))
+							continue;
+						rotations = newRotations;
+						pcursmonst = mi;
+						maxSteps = node.steps;

```

Comment:
Here it looks like we've found the closest monster, shouldn't we abort the search?